### PR TITLE
Add benchmarks for scaled_mm and scaled_grouped_mm

### PR DESCRIPTION
**Pull request by jainapurva**
### Summary
Add operator benchmarks for scaled_mm and scaled_grouped_mm operations
This PR introduces comprehensive microbenchmarks for FP8 scaled matrix multiplication operations, targeting large language model inference and training workloads.

### What's Added
**New file:** benchmarks/operator_benchmark/pt/scaled_mm_test.py; benchmarks/operator_benchmark/pt/scaled_grouped_mm_test.py
**Two benchmark classes:**
1. ScaledMMBenchmark - Tests torch.nn.functional.scaled_mm for standard FP8 matrix multiplication
2. ScaledGroupedMMBenchmark - Tests torch.nn.functional.scaled_grouped_mm for MoE (Mixture of Experts) grouped matmul operations

### Model Shapes
Benchmarks cover realistic production shapes from two major model architectures:
- **DeepSeek-V3 671B:** Matrix dimensions (M, N=2048, K=7168)
- **Llama4 16e:** Matrix dimensions (M, N=8192, K=5120)

With varying batch sizes (M=16384, 128000) and expert counts (G=1,2,4,8) for grouped operations.

### CODE REVIEW

File: .github/workflows/operator_microbenchmark.yml [hunk 1]
```diff
@@ -37,8 +37,8 @@ jobs:
     needs: get-label-type
     with:
       runner: linux.12xlarge.memory
-      build-environment: linux-jammy-cuda12.8-py3.10-gcc11-sm80
-      docker-image-name: ci-image:pytorch-linux-jammy-cuda12.8-cudnn9-py3-gcc11
+      build-environment: linux-jammy-cuda12.9-py3.10-gcc11-sm80
```
Comment (slayton58) [inline .github/workflows/operator_microbenchmark.yml]: Intended?

File: benchmarks/operator_benchmark/pt/scaled_grouped_mm_test.py [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+    Reuse scale/quantization helpers from `test/test_scaled_matmul_cuda.py`.
+
+    `test/` isn't a package, so we import by path and cache the module.
+    """
+    global _TEST_SCALED_MATMUL_CUDA_MOD
+    if _TEST_SCALED_MATMUL_CUDA_MOD is not None:
+        return _TEST_SCALED_MATMUL_CUDA_MOD
+
+    pytorch_root = os.path.abspath(
+        os.path.join(os.path.dirname(__file__), "..", "..", "..")
```
Comment (slayton58) [inline benchmarks/operator_benchmark/pt/scaled_grouped_mm_test.py]: This is to grab the helper functions from the test files?

File: benchmarks/operator_benchmark/pt/scaled_grouped_mm_test.py [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+            raise ValueError(f"Unsupported scaling format: {scaling}")
+
+        self.set_module_name("scaled_grouped_mm")
+
+    def _init_fp8(self, M, N, K, G, device, float8_dtype, scaling):
+        """Initialize FP8 tensorwise or rowwise scaling."""
+        self.float8_dtype = _get_float8_dtype(float8_dtype)
+
+        # We interpret offs as group end offsets along K (grouped-K).
+        # Use deterministic equal-sized groups.
```
Comment (slayton58) [inline benchmarks/operator_benchmark/pt/scaled_grouped_mm_test.py:101]: Do we want an option to use non-equal groups?

File: benchmarks/operator_benchmark/pt/scaled_grouped_mm_test.py [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+    (128000, 8192, 5120, 4),
+    (128000, 8192, 5120, 8),
+    (16384, 1536, 5120, 1),
+    (16384, 1536, 5120, 8),
+    (128000, 1536, 5120, 1),
+    (128000, 1536, 5120, 8),
+    (16384, 2048, 7168, 1),
+    (16384, 2048, 7168, 8),
+    (128000, 2048, 7168, 1),
+    (128000, 2048, 7168, 8),
```
Comment (slayton58) [inline benchmarks/operator_benchmark/pt/scaled_grouped_mm_test.py]: @danielvegamyhre Are these reasonable problem sizes to bench?

File: benchmarks/operator_benchmark/pt/scaled_mm_test.py [hunk 5]
```diff
@@ -0,0 +1,597 @@
...
+    """
+    Reuse scale/quantization helpers from `test/test_scaled_matmul_cuda.py`.
+    `test/` isn't a package, so we import by path and cache the module.
+    """
+    global _TEST_SCALED_MATMUL_CUDA_MOD
+    if _TEST_SCALED_MATMUL_CUDA_MOD is not None:
+        return _TEST_SCALED_MATMUL_CUDA_MOD
+
+    pytorch_root = os.path.abspath(
+        os.path.join(os.path.dirname(__file__), "..", "..", "..")
```
Comment (slayton58) [inline benchmarks/operator_benchmark/pt/scaled_mm_test.py]: DRY from the grouped file

File: benchmarks/operator_benchmark/pt/scaled_mm_test.py [hunk 6]
```diff
@@ -0,0 +1,597 @@
...
+    if spec is None or spec.loader is None:
+        raise RuntimeError(f"Failed to create import spec for {test_file}")
+
+    mod = importlib.util.module_from_spec(spec)
+    spec.loader.exec_module(mod)
+    _TEST_SCALED_MATMUL_CUDA_MOD = mod
+    return mod
+
+
+def _get_float8_dtype(float8_dtype):
```
Comment (slayton58) [inline benchmarks/operator_benchmark/pt/scaled_mm_test.py]: DRY again

File: benchmarks/operator_benchmark/pt/scaled_mm_test.py [hunk 7]
```diff
@@ -0,0 +1,597 @@
...
+            "y": y_lp.t(),  # mat_b is (K, N)
+            "scale_a": x_scales.reciprocal(),
+            "scale_b": y_scales.reciprocal(),
+        }
+        self._set_scaled_mm_call_config(
+            scale_recipe_a=ScalingType.BlockWise1x128,
+            scale_recipe_b=ScalingType.BlockWise1x128,
+        )
+
+    def _init_fp8_blockwise_128x128(
```
Comment (slayton58) [inline benchmarks/operator_benchmark/pt/scaled_mm_test.py]: I feel like this and `_init_fp8_blockwise_1x128` are incredibly similar - refactor to minimize code duplication?

File: benchmarks/operator_benchmark/pt/scaled_mm_test.py [hunk 8]
```diff
@@ -0,0 +1,597 @@
...
+            scale_recipe_b=ScalingType.BlockWise128x128,
+        )
+
+    def _init_mx_blockwise(
+        self, M: int, N: int, K: int, device: str, *, mx_format: str
+    ) -> None:
+        # MX uses BlockWise1x32 with swizzled scales on CUDA.
+        if device != "cuda":
+            raise RuntimeError(f"MX scaling requires CUDA device, got: {device}")
+        if torch.version.hip is not None:
```
Comment (slayton58) [inline benchmarks/operator_benchmark/pt/scaled_mm_test.py]: I think this is pretty simple to fix - just use `SwizzleType.NO_SWIZZLE` below for AMD, but if we're not benchmarking on AMD right now, then maybe it just doesn't matter :)

File: benchmarks/operator_benchmark/pt/scaled_mm_test.py [hunk 9]
```diff
@@ -0,0 +1,597 @@
...
+        }
+        if self._swizzle_a is not None:
+            kwargs["swizzle_a"] = self._swizzle_a
+        if self._swizzle_b is not None:
+            kwargs["swizzle_b"] = self._swizzle_b
+
+        return torch.nn.functional.scaled_mm(x, y, **kwargs)
+
+
+MNK_list = [
```
Comment (slayton58) [inline benchmarks/operator_benchmark/pt/scaled_mm_test.py]: Do we want some smaller tests too? I.e. M < 128 and the like, to cover non massive training cases

File: benchmarks/operator_benchmark/pt/scaled_mm_test.py [hunk 10]
```diff
@@ -0,0 +1,597 @@
...
+                "device": ["cuda"],
+                "float8_dtype": ["e4m3fn"],
+                "output_dtype": rowwise_output_dtypes,
+                "scaling": ["fp8_rowwise"],
+            },
+            tags=["long"],
+        )
+
+    # MX + NVFP4: keep bf16-only for now.
+    # These are CUDA-only (non-HIP) due to swizzled scale requirements.
```
Comment (slayton58) [inline benchmarks/operator_benchmark/pt/scaled_mm_test.py]: See above - MXFP4 has HIP implementation w/non-swizzled scales, which should be a fairly simple change

**Review by slayton58 — APPROVED** [review]
LGTM - few nits, nothing major

File: benchmarks/operator_benchmark/pt/scaled_grouped_mm_test.py (hunk 2 above)
Comment (jainapurva) [reply benchmarks/operator_benchmark/pt/scaled_grouped_mm_test.py]: Yes, As the test file is not a module, had to get a little creative for it

File: .github/workflows/operator_microbenchmark.yml (hunk 1 above)
Comment (jainapurva) [reply .github/workflows/operator_microbenchmark.yml]: Yes, there's a cutlass error if i build torch with cu128 on H100 for fp8 rowwise benchmarks.

File: benchmarks/operator_benchmark/pt/scaled_grouped_mm_test.py (hunk 4 above)
Comment (danielvegamyhre) [reply benchmarks/operator_benchmark/pt/scaled_grouped_mm_test.py]: sorry for the insanely late reply, not sure how i missed this, will check my notifications for pytorch/pytorch - but yes, these shapes are fine.

looks like dsv3 671b and llama4 shapes for routed expert dim/hidden_dim, and for the leading/batch dim, 32k-128k or so is what i generally test. 

for MoE training people often use dp2ep parallelism and try to have a large "total M" dim to improve GEMM efficiency with these small/granular experts used by the dsv3/qwen/kimi/etc style models. so i focus on that.
